### ISSUE TITLE
Syntax highlighting failure in GVim on Windows

### ISSUE THREAD

**Issue by skywind3000** [vim/vim#1918]
GVim version: 8.0.839 x86
Windows: Win 10 x64

GVim 8.0.636 works fine, but all the highlighting disappeared when I move around the cursor in GVim 8.0.839. Console mode is also fine, only the GVim:

![syntax_error](https://user-images.githubusercontent.com/3035071/28892652-48fa7120-7801-11e7-8bf1-f6719c8db877.gif)

command to reproduce:

```bash
gvim -N -u ..\initrc.vim d:/acm/github/vim/asc/menu.vim
```

content of **initrc.vim**:

```VimL
set nocompatible
syntax on
```

If I enter these two line above in the vim, the bug will not be reproduced, they must be in initrc.vim and load before gvim startup.

**Comment by markonm** [comment]
I can reproduce similar Vim script syntax fail on 8.0.1067 on Win 10 x64. In my case syntax breaks in gVim and CMD's Vim.

Steps to reproduce:

1. `gvim -u NONE -i NONE -N "+so $VIMRUNTIME/plugin/matchparen.vim" "+syntax on" reproduce.vim`
2. type the following in normal mode `ffyiWjviwpk`
3. syntax disappears

reproduce.vim
```vim
let a = foo()
let b = baz
```

If I disable `matchparen` with `let g:loaded_matchparen  = 1`, syntax will not break

Breakage was introduced in [patch 8.0.0647](https://github.com/vim/vim/commit/06f1ed2f78c5c03af95054fc3a8665df39dec362). Before that commit I cannot reproduce.

**Comment by skywind3000** [comment]
Still exists in 8.0.1118

**Comment by mattn** [comment]
@xtal8 I also use Win 10 x64. But I don't reproduce it.

**Comment by markonm** [comment]
@mattn Do you use x64 gVim also? I can't reproduce on x86. My version is downloaded from here https://github.com/vim/vim-win32-installer/releases

**Comment by mattn** [comment]
Yes, I use x64. But I built vim with mingw compiler.

**Comment by markonm** [comment]
Could you try to reproduce with  gvim_8.0.1118_x64.zip from that link?

**Comment by skywind3000** [comment]
I always download gvim from 
https://github.com/vim/vim-win32-installer/releases

And both gvim_8.0.1118_x64.zip and gvim_8.0.1118_x86.zip downloaded from 
can reproduce this problem:

1. download my menu.vim from 
https://github.com/skywind3000/vim/blob/master/asc/menu.vim

2. open menu.vim and 

```VimL
:normal gg$
```

3. move the cursor up and down by 'j' and 'k'

and than, You can encounter the issue in both gvim.exe and vim.exe.

attachment: menu.vim
[menu.zip](https://github.com/vim/vim/files/1308878/menu.zip)

**Comment by markonm** [comment]
I can reproduce issue with menu.vim only on gvim_8.0.1118_x64.zip. When I try to reproduce on x86. there's no syntax breakage.

**Comment by skywind3000** [comment]
strange, on my laptop:

gvim installed by gvim_8.0.1118_x86.exe has the problem.
gvim extracted from gvim_8.0.1118_x86.zip doesn't have the problem.

**Comment by chrisbra** [comment]
> strange, on my laptop:
> 
> gvim installed by gvim_8.0.1118_x86.exe has the problem.
> gvim extracted from gvim_8.0.1118_x86.zip doesn't have the problem.

Just checked it, could not reproduce it with either one.

**Comment by mattn** [comment]
This patch will fix the issue.

```diff
diff --git a/src/syntax.c b/src/syntax.c
index 6361bd7aa..0d5af0cc3 100644
--- a/src/syntax.c
+++ b/src/syntax.c
@@ -6573,9 +6573,14 @@ syn_get_id(
     else if (wp->w_buffer == syn_buf
 	    && lnum == current_lnum
 	    && col > current_col)
+    {
+	/* reset syn_tm here since this is not highlighting */
+	syn_tm = NULL;
+
 	/* next_match may not be correct when moving around, e.g. with the
 	 * "skip" expression in searchpair() */
 	next_match_idx = -1;
+    }
 
     (void)get_syntax_attr(col, spellp, keep_state);
 
```
syn_tm is set when start syntax to disable heavy syntax. matchparen call f_synstack(). And it call syn_get_id(), syn_get_attr(). This doesn't call syntax_call (this is correct pass). i.e. syn_tm is used unintentionally which is already timed-out.

**Comment by mattn** [comment]
@brammool This should be temporary updating like below?

```diff
diff --git a/src/syntax.c b/src/syntax.c
index 6361bd7aa..950ea7153 100644
--- a/src/syntax.c
+++ b/src/syntax.c
@@ -6564,6 +6564,8 @@ syn_get_id(
     int		*spellp,     /* return: can do spell checking */
     int		keep_state)  /* keep state of char at "col" */
 {
+    proftime_T	*save_tm = syn_tm;
+
     /* When the position is not after the current position and in the same
      * line of the same buffer, need to restart parsing. */
     if (wp->w_buffer != syn_buf
@@ -6573,12 +6575,19 @@ syn_get_id(
     else if (wp->w_buffer == syn_buf
 	    && lnum == current_lnum
 	    && col > current_col)
+    {
+	/* reset syn_tm here since this is not highlighting */
+	syn_tm = NULL;
+
 	/* next_match may not be correct when moving around, e.g. with the
 	 * "skip" expression in searchpair() */
 	next_match_idx = -1;
+    }
 
     (void)get_syntax_attr(col, spellp, keep_state);
 
+    syn_tm = save_tm;
+
     return (trans ? current_trans_id : current_id);
 }
 
```

**Comment by brammool** [comment]
Fixed by 8.0.1133
